test/parallel_api/experimental/transform_async_binary_zip.pass.cpp

```diff
@@ -0,0 +1,181 @@
+// -*- C++ -*-
+//===-- transform_binary.pass.cpp -----------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#include "support/test_config.h"
+
+#include _PSTL_TEST_HEADER(execution)
+#include _PSTL_TEST_HEADER(algorithm)
+
+#if TEST_DPCPP_BACKEND_PRESENT
+#    include "oneapi/dpl/async"
+#endif // TEST_DPCPP_BACKEND_PRESENT
+
+#include "support/utils.h"
+
+using namespace TestUtils;
+
+#if TEST_DPCPP_BACKEND_PRESENT
+template <typename TInput1, typename TInput2, typename TOutput>
+class TheOperation
+{
+    TOutput val;
+
+public:
+
+    TheOperation(TOutput v) : val(v) {}
+
+    TOutput
+    operator()(const ::std::tuple<const TInput1&, const TInput1&>& x,
+               const ::std::tuple<const TInput2&, const TInput2&>& y) const
+    {
+        return TOutput(val + (::std::get<0>(x) + ::std::get<1>(x)) - (::std::get<0>(y) + ::std::get<1>(y)));
+    }
+};
+
+template <typename InputIterator1, typename InputIterator2, typename OutputIterator>
+void
+check_and_reset(InputIterator1 first1, InputIterator1 last1, InputIterator2 first2, OutputIterator out_first)
+{
+    typedef typename ::std::iterator_traits<OutputIterator>::value_type Out;
+    typename ::std::iterator_traits<OutputIterator>::difference_type k = 0;
+    for (; first1 != last1; ++first1, ++first2, ++out_first, ++k)
+    {
+        // check
+        Out expected = Out(1.5) + (::std::get<0>(*first1) + ::std::get<1>(*first1)) - (::std::get<0>(*first2) + ::std::get<1>(*first2));
+        Out actual = *out_first;
+        if (::std::is_floating_point_v<Out>)
+        {
+            EXPECT_TRUE((expected > actual ? expected - actual : actual - expected) < Out(1e-7),
+                        "wrong value in output sequence");
+        }
+        else
+        {
+            EXPECT_EQ(expected, actual, "wrong value in output sequence");
+        }
+        // reset
+        *out_first = k % 7 != 4 ? 7 * k + 5 : 0;
+    }
+}
+
+template <typename T1, typename T2, typename T3>
+struct test_one_policy
+{
+    template <typename Policy, typename InputIterator1, typename InputIterator2, typename OutputIterator,
+              typename BinaryOp>
+    void
+    operator()(Policy&& exec, InputIterator1 first1, InputIterator1 last1, InputIterator2 first2, InputIterator2 /* last2 */,
+               OutputIterator out_first, OutputIterator /* out_last */, BinaryOp op)
+    {
+        auto result = oneapi::dpl::experimental::transform_async(exec, first1, last1, first2, out_first, op);
```

Comment:
Do I understand correctly that we have just one difference with call `oneapi::dpl::experimental::transform_async` instead of `::std::transform` (`dpl::transform`) comparing with file `transform_binary_zip.pass.cpp` ?

Reply:
No, it's incorrect understanding: also predicate in these new tests working with `std::tuple`

Reply:
Please have a look at the "diff" for these files..
[diff_transform_binary_zip_transform_async_binary_zip.diff.txt](https://github.com/oneapi-src/oneDPL/files/13979375/diff_transform_binary_zip_transform_async_binary_zip.diff.txt)

Essentially, I would say there is no significant difference expecting "wait()" and test invocation  just for hetero policies in case of `transform_async`.

To tell the truth I have an objection here. I don't like huge code duplication, because:
1) A reviewer has to review the same code twice.
2) A reviewer has to leave a feedback at the same places twice.
3) An author has to fix the found error twice.
4) In case of refactoring/modification/improving the duplicated code, a developer has to don't  miss the duplicated places to modify ones as well, and has to  write the same new code twice.

If the author doesn't know how how to get rid of the code duplication, please let me know - I will help with that.
If the author  is strongly disagree with my point of view regarding the code duplication, let's  make a voting :) If I am in the minority, I will follow "disagree and commit".

Reply:
@MikeDvorskiy could you please help with this point?

Reply:
> @MikeDvorskiy could you please help with this point?

sure, but let me to think a little bit..

Reply:
@SergeyKopienko, you may review the new commits within this PR.